rviz_tactile_plugins/src/tactile_visual_base.cpp

```diff
@@ -106,8 +109,13 @@ float TactileVisualBase::mapValue(const ::tactile::TactileValue &value)
   // normalize to range 0..1
   if (mode_ == ::tactile::TactileValue::rawCurrent ||
       mode_ == ::tactile::TactileValue::rawMean) {
-    v = (v - raw_range_.min()) / raw_range_.range();
+    if (raw_range_.range() != 0.0 ) { // if fixed range, and manually set to same value, range could be 0
+      v = (v - raw_range_.min()) / raw_range_.range();
+    }
   }
+  // clamp to 0..1 (permits to accept fixed range and saturate outside)
+  if (v > 1.0) v = 1.0;
+  if (v < 0.0) v = 0.0;
```

Comment:
The clamping is actually already handled in `mapColor()`. So why did you added this here? I cannot imagine that you observed any issues without that.
I thought about some visual indication that the values were clamped. As we already have pink for an error state, we could use some dark pink for lower and some "redish" pink for upper clamp. What do you think?

Reply:
The existing code of the section you refer to, was providing a [0..1] range in any case, so to not break this assumption (and not rely on external mechanisms), I think I added the clamping to be fully backward compatible.

I know that `mapColor` clamps the colors [here](https://github.com/ubi-agni/tactile_toolbox/blob/melodic-devel/rviz_tactile_plugins/src/color_map.cpp#L69)  but `mapValues` is used at other places, for example to display arrows if enable (I already used this debug feature) [here](https://github.com/ubi-agni/tactile_toolbox/blob/melodic-devel/rviz_tactile_plugins/src/tactile_taxels_visual.cpp#L240)

I don't think the color of the taxels should be affected by the saturation. In the example scenario such a color jump was not desirable at least.
One could maybe change the font color of the range values in the range property, to indicate that the range::update function would like to update the max but it is currently "locked".

Reply:
> I don't think the color of the taxels should be affected by the saturation. In the example scenario such a color jump was not desirable at least.

It will not be a color jump, but a smooth transition into the error color (or half way into it as I designed it now).
I think it's important to see that the values exceed the bounds. A red font of the range values doesn't help very much, because the range values are global for all sensors, but the color indication should be per taxel.

> `mapValues` is used at other places, for example to display arrows if enabled

Actually I planned to remove those arrows. Isn't that handled by the contact state display (and your new no-clustering mode)?

Reply:
a smooth transition is even worse in my opinion. Will we have then a transition from red to redish pink ? 

If you insist on displaying the saturation on the taxel, I think saturation should be displayed at the max not before, so nothing smooth should happen before the saturation occurs. 
I still would prefer no color change. Is there a possibility to draw a duplicate of the taxel mesh as a thin wireframe object instead of the filled object and then have the wireframe in red ? ideally only contour (not inner triangle of the mesh) but I think this becomes really hard to do)

or just add a red primitive object on top of the taxel (at the center), even a text ("max") would be fine for me, but a transition is wrong to me. saturation is saturation. it should stay red at the max and black at the min. 

Moreover, I designed the lock range feature to make video shots for a colleague and he wanted the full range of color with no extra display. Indicating saturation would be wrong for his application, as we were reaching saturation immediately (hand baro).
If this feature is not matching his needs, I will have to keep my PR on a separate fork for him I suppose.

I suppose there is no reason to discuss anymore the clamp. I matched the previous behavior of the function in my PR and you now want a new behavior (letting the value unclamped and clamping in the consumer of the value). it is a design choice that I have no opinion on so don't ask me more. you have my reasons why I did that.

Regarding the display of the arrows, I find it practical, because I did not need to start the second application to display the direction and centers, but we indeed have a similar function now in the `tactile_merger`